beacon-chain/forkchoice/protoarray/node.go

```diff
@@ -34,17 +48,125 @@ func (n *Node) Weight() uint64 {
+		return ctx.Err()
+	}
+	// update the children
+	childrensWeight := uint64(0)
+	for _, child := range n.children {
+		if err := child.applyWeightChanges(ctx); err != nil {
+			return err
+		}
+		childrensWeight += child.weight
+	}
+	if n.root == params.BeaconConfig().ZeroHash {
+		return nil
+	}
+	n.weight = n.balance + childrensWeight
+	return nil
+}
+
+// updateBestDescendant updates the best descendant of this node and its
+// children.
+func (n *Node) updateBestDescendant(ctx context.Context, justifiedEpoch, finalizedEpoch types.Epoch) error {
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+	if len(n.children) == 0 {
+		n.bestDescendant = nil
+		return nil
+	}
+
+	var bestChild *Node
+	bestWeight := uint64(0)
+	hasViableDescendant := false
+	for _, child := range n.children {
+		if child == nil {
+			return errNilNode
+		}
+		if err := child.updateBestDescendant(ctx, justifiedEpoch, finalizedEpoch); err != nil {
+			return err
+		}
+		childLeadsToViableHead := child.leadsToViableHead(justifiedEpoch, finalizedEpoch)
+		if childLeadsToViableHead && !hasViableDescendant {
+			// The child leads to a viable head, but the current parent's best child doesnt.
+			bestWeight = child.weight
+			bestChild = child
+			hasViableDescendant = true
+		} else if childLeadsToViableHead {
+			// If both are viable, compare their weights.
+			if child.weight == bestWeight {
+				// Tie-breaker of equal weights by root.
+				if bytes.Compare(child.root[:], bestChild.root[:]) > 0 {
+					bestChild = child
+				}
+			} else if child.weight > bestWeight {
+				bestChild = child
+				bestWeight = child.weight
+			}
+		}
+	}
+	if hasViableDescendant {
+		if bestChild.bestDescendant == nil {
+			n.bestDescendant = bestChild
+		} else {
+			n.bestDescendant = bestChild.bestDescendant
+		}
+	} else {
+		n.bestDescendant = nil
+	}
+	return nil
+}
+
+// viableForHead returns true if the node is viable to head.
+// Any node with diff finalized or justified epoch than the ones in fork choice store
+// should not be viable to head.
+func (n *Node) viableForHead(justifiedEpoch, finalizedEpoch types.Epoch) bool {
+	justified := justifiedEpoch == n.justifiedEpoch || justifiedEpoch == 0
+	finalized := finalizedEpoch == n.finalizedEpoch || finalizedEpoch == 0
+
+	return justified && finalized
+}
+
+func (n *Node) leadsToViableHead(justifiedEpoch, finalizedEpoch types.Epoch) bool {
+	if n.bestDescendant == nil {
+		return n.viableForHead(justifiedEpoch, finalizedEpoch)
+	}
+	return n.bestDescendant.viableForHead(justifiedEpoch, finalizedEpoch)
 }
 
 // BestDescendant of the fork choice node.
-func (n *Node) BestDescendant() uint64 {
+func (n *Node) BestDescendant() *Node {
 	return n.bestDescendant
 }
 
-// Graffiti of the fork choice node.
-func (n *Node) Graffiti() [32]byte {
-	return n.graffiti
+// setFullyValidated sets the current node and all of its ancestors as fully
+// validated (ie. non-optimistic) nodes
+func (n *Node) setFullyValidated(ctx context.Context) error {
```

Comment:
This function doesn't do anything other than return `nil`

Reply:
Oh very nice catch, I had tests for optimistic status, thought they covered this, adding a specific unit test